File: apps/web/src/server/rate-limit.ts
import { cacheScope } from "./upstash";

type Bucket = { count: number; reset: number };

const memory = new Map<string, Bucket>();

function windowMs() {
  return Number(process.env.RATE_LIMIT_WINDOW_MS ?? 60_000);
}

function limitFor(kind: "auth" | "chat" | "connect") {
  if (kind === "auth") return Number(process.env.RATE_LIMIT_AUTH ?? 20);
  if (kind === "connect") return Number(process.env.RATE_LIMIT_CONNECT ?? 30);
  return Number(process.env.RATE_LIMIT_CHAT ?? 120);
}

function blocked(reset: number) {
  return new Response(JSON.stringify({ detail: "Rate limit exceeded. Try again shortly." }), {
    status: 429,
    headers: {
      "content-type": "application/json",
      "retry-after": String(Math.max(1, Math.ceil((reset - Date.now()) / 1000))),
    },
  });
}

/**
 * Fixed-window rate limit. Uses atomic INCR when Upstash REST is configured.
 */
export async function rateLimit(
  kind: "auth" | "chat" | "connect",
  identity: string,
): Promise<Response | null> {
  const id = cacheScope(identity || "anon");
  const limit = limitFor(kind);
  const window = windowMs();
  const bucketKey = `pm:rl:${kind}:${id}`;
  const now = Date.now();

  const url = process.env.UPSTASH_REDIS_REST_URL?.trim();
  const token = process.env.UPSTASH_REDIS_REST_TOKEN?.trim();
  if (url && token) {
    try {
      const incr = await fetch(`${url}/incr/${encodeURIComponent(bucketKey)}`, {
        headers: { Authorization: `Bearer ${token}` },
      });
      const body = (await incr.json()) as { result?: number };
      const count = Number(body.result ?? 0);
      if (count === 1) {
        await fetch(`${url}/pexpire/${encodeURIComponent(bucketKey)}/${window}`, {
          headers: { Authorization: `Bearer ${token}` },
        });
      }
      if (count > limit) {
        return blocked(now + window);
      }
      return null;
    } catch {
      /* fall through to memory */
    }
  }

  const cur = memory.get(bucketKey) ?? { count: 0, reset: now + window };
  if (cur.reset < now) {
    cur.count = 0;
    cur.reset = now + window;
  }
  cur.count += 1;
  memory.set(bucketKey, cur);
  if (cur.count > limit) return blocked(cur.reset);
  return null;
}

export function clientIp(request: Request): string {
  const xf = request.headers.get("x-forwarded-for");
  if (xf) return xf.split(",")[0]?.trim() || "anon";
  return request.headers.get("x-real-ip") || "anon";
}

File: apps/api/app/core/rate_limit.py
from __future__ import annotations

import time
from typing import Literal

from fastapi import HTTPException, Request

Kind = Literal["chat", "connect", "auth"]

_memory: dict[str, dict[str, float | int]] = {}

_LIMITS: dict[Kind, int] = {
    "chat": 120,
    "connect": 30,
    "auth": 20,
}
_WINDOW_MS = 60_000


def _client_ip(request: Request) -> str:
    xf = request.headers.get("x-forwarded-for")
    if xf:
        return xf.split(",")[0].strip() or "anon"
    if request.client:
        return request.client.host or "anon"
    return "anon"


def check_rate_limit(kind: Kind, identity: str) -> None:
    """Fixed-window in-memory rate limit. Raises 429 when exceeded."""
    limit = _LIMITS[kind]
    now = time.time() * 1000
    key = f"pm:rl:{kind}:{identity or 'anon'}"
    cur = _memory.get(key)
    if cur is None or float(cur["reset"]) < now:
        cur = {"count": 0, "reset": now + _WINDOW_MS}
        _memory[key] = cur
    cur["count"] = int(cur["count"]) + 1
    if int(cur["count"]) > limit:
        retry = max(1, int((float(cur["reset"]) - now) / 1000))
        raise HTTPException(
            status_code=429,
            detail="Rate limit exceeded. Try again shortly.",
            headers={"Retry-After": str(retry)},
        )


async def rate_limit_chat(request: Request) -> None:
    check_rate_limit("chat", _client_ip(request))


async def rate_limit_connect(request: Request) -> None:
    check_rate_limit("connect", _client_ip(request))
